File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        """
        dfs takes a path array and results array
            if the length of the path equals the length of the digits
                we've come up with a full permutation and can append it to the res arr and return
            
            next number is the index of digits equal to the current length of path 
            e.g. digits[1] for the 2nd number if path only has the 1st number, due to zero indexing
            
            iterate over the letters in this next number
                append it to the path
                call this recursive function on it
                after we recurse to the end of the permutation, we pop it off the path
        """
        KEYBOARD = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz',
        }
        if digits == "": return []
        def dfs(path, res):
            if len(path) == len(digits):
                res.append(''.join(path))
                return
            
            next_number = digits[len(path)]
            for letter in KEYBOARD[next_number]:
                path.append(letter)
                dfs(path, res)
                path.pop()
        
        res = []
        dfs([], res)
        return res
    
    

File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    
    private Map<Character, String> map = 
        Map.of('2', "abc", '3', "def", '4', "ghi", '5', "jkl", '6', "mno", '7', "pqrs", '8', "tuv", '9', "wxyz");
    
    private List<String> res = new LinkedList<String>();
    
    private String phoneDigits;
    
    public List<String> letterCombinations(String digits) {
        
        if (digits.length() == 0) return res;
        phoneDigits = digits;
        dfs(new StringBuilder(), 0);
        return res;        
    }
    
    private void dfs(StringBuilder path, int i){
        if (path.length() == phoneDigits.length()){
            res.add(path.toString());
            return;
        }
        
        String digitLetters = map.get(phoneDigits.charAt(i));
        for (char letter : digitLetters.toCharArray()){
            path.append(letter);
            dfs(path, i+1);
            path.deleteCharAt(path.length()-1);
            
        }
    }
}